### PR TITLE
fix deadlock in model.train mode forward run only

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: fix deadlock in model.train model forward run only.

Issues identified and investigated by Aishwarya. 

When training program exits, PyNodeSharedPointerPool destructor is called, if grad_fns_ is not empty,
PyNode::release_variables() will be called (https://github.com/pytorch/pytorch/blob/15532595209d2daf34d35e10f8d3d3b64966aea2/torch/csrc/autograd/python_function.cpp#L168)

The other hand, there is known issue when acquiring GIL in pybind11 destructors, there will be probabbly deadlock issue (https://github.com/pybind/pybind11/issues/1446)

The resolution here, we remove all maintained states before ~PyNodeSharedPointerPool happens (e.g. the python program exits) to avoid the deadlock issue.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by wschin** [comment]
Just my two cents. For a context, it's attached to a Pytorch tensor. Could we release context when the associated tensor is deleted? Maybe we need to record both tensor and its context in FW and BW runner. At the beginning of FW and BW, we scan all recorded tensors and delete unused contexts. Would it be better than caching all contexts to the end of training script?

**Review by ganik — DISMISSED** [review]

**Comment by pengwa** [comment]
> 

without the PyNodeSharedPointerPool, context will be released as soon as the tensor is not used by any other nodes. The problem is, if the ctx is used in backward, at a time, the ctx and tensor are already released, we will see segment fault. 

What I think better here is, we clear the pool before each forward run, which somehow make senses and aligned with torch behavior. Imagine if a torch model runs forward only, and once the loss tensor is out of scope next time, the ctx got removed. And usually the time a loss tensor become out of scope happens when next loss is calculated, and variable got overridden. We can talk tomorrow for more details. :)

**Review by ganik — DISMISSED** [review]

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/torch_interop_utils/torch_interop_utils.cc [hunk 1]
```diff
@@ -109,8 +114,20 @@ void unregister_grad_fn(size_t ctx_address)
   PyNodeSharedPointerPool::GetInstance().UnRegisterGradFunc(ctx_address);
 }
 
+// Supposed to be cleared on python program exit or before every forward run to resolve following bug:
+// When training program exits, PyNodeSharedPointerPool destructor is called, if grad_fns_ is not empty,
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/torch_interop_utils/torch_interop_utils.cc:118]: It seems like the `PyNodeSharedPointerPool` destructor is private. Where and how does it get called?
Who is invoking `release_variables`? Are we able to somehow use some signal to automatically clear `grad_fns_` when `release_variables` is called?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 2]
```diff
@@ -39,6 +40,10 @@ def __init__(self, model, debug_options: DebugOptions, fallback_manager: _Fallba
     def execution_session_run_forward(execution_session, onnx_model, device, gradient_accumulation_manager, *inputs):
         """Runs the forward graph on execution_session with given model inputs and device"""
 
+        # Clear all gradient functions, to avoid a deadlock issue.
+        # Check the called function for more detailed comments.
+        clear_all_grad_fns()
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:45]: in this case, there will not be deadlock, but there are memory leak because we accumulate grad_fns without releasing them (which happens in backward)

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/torch_interop_utils/torch_interop_utils.cc (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/torch_interop_utils/torch_interop_utils.cc:118]: PyNode::release_variables() is called by PyTorch when  the auto gradient graph got destructed (for example, backward run completed or loss function out of its lifecycle). 

Since release_variables is called when some torch Node tears down when the Node's ref cnt being 0 among the overall autograd graph, so we actually don't have a easy way to do what you suggested here.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py [hunk 3]
```diff
@@ -25,13 +25,17 @@ def enable_custom_autograd_support():
...
     import atexit
 
     register_forward_runner(call_python_forward_function)
     register_backward_runner(call_python_backward_function)
 
     # Unregister all python functions automatically upon normal interpreter termination.
     atexit.register(unregister_python_functions)
+    # Clear all gradient functions, to avoid a deadlock issue.
+    # Check the called function for more detailed comments.
+    atexit.register(torch_interop_utils.clear_all_grad_fns)
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py:38]: Oh yeah, you are right, we can move this into torch_interop_utils module. 

For the second question, I did not find such an approach acting as C++ destructors corresponding to __init__ constructor. 
In terms of RAII, do you mean we still add python context covering ORTModule forward + backward run? I believe this is not feasible? Feel free to correct me if I am wrong.

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/torch_interop_utils/torch_interop_utils.cc (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/torch_interop_utils/torch_interop_utils.cc:118]: In ORT, PythonOpGrad are responsible to release the shared pointer of corresponding torch context object, which will trigger `release_variables`. But we don't have a way to handle the releasing if backward pass is not called.

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:45]: updated the comments.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py:38]: No worries. I was just hoping that we could somehow auto destruct the object based on some signal. Seems like it might be not straightforward. I think this is good for now.

**Review by baijumeswani — APPROVED** [review]
